revamp the way routes are defined in app resources

jcarrothers-sap opened the pull request:
- the "default_route" from the old "route" block is the only entry still allowed
- The new "routes" block is similar to the "service_bindings" block in that
  it is a set of maps and each map defines a single route mapping.

fixes #62

gberche-orange commented:
thanks @jcarrothers-sap 

We should consider applying the [Deprecations, Removals, and Renames best practice](https://www.terraform.io/docs/extend/best-practices/deprecations.html), especially if we release 1.0 without this PR.

jcarrothers-sap commented:
@gberche-orange Agreed, I specifically did flag any fields I thought would actually be in use a removed or deprecated, but the fact that this is a big shift in the way routes are defined is why I think it would be best if this got into the 1.0 release.

jcarrothers-sap commented on cloudfoundry/resource_cf_app.go:
```diff
@@ -254,9 +260,32 @@ func resourceApp() *schema.Resource {
 							Type:     schema.TypeString,
 							Optional: true,
 						},
-						"version": &schema.Schema{
```
I simply removed this field due to not expecting it to be in use by anyone since it has never had any implemented functionality.  Also, it is not currently documented.

jcarrothers-sap commented on cloudfoundry/resource_cf_app.go:
```diff
@@ -254,9 +260,32 @@ func resourceApp() *schema.Resource {
 							Type:     schema.TypeString,
 							Optional: true,
```
Looking at this again, I think it would be best to flag this `validation_script` field as removed as well.  It's removal shouldn't be a problem as it wasn't implemented anyways. Also, it is not currently documented.

gberche-orange replied:
+1

jcarrothers-sap commented on website/docs/r/app.html.markdown:
```diff
@@ -87,8 +87,10 @@ One of the following arguments must be declared to locate application source or
+* `routes` - (Optional, Array) Configures how the application will be accessed externally to cloudfoundry. 
+
+  - `route` - (Required, String) The route GUID.
+  - `port` - (Optional, Number) The port of the application that the route mapping is to bind to (NOT YET IMPLEMENTED!)
```
I think that makes sense, we can always add it later if needed.  I'll update the documentation, but I'm going to leave the property in the code for now as it is an attribute that is populated once an app has been created.

jcarrothers-sap commented on cloudfoundry/resource_cf_app.go:
```diff
@@ -999,7 +1211,22 @@ func prepareApp(app cfapi.CCApp, d *schema.ResourceData, log *cfapi.Logger) (pat
+			}
+		}
+		return fmt.Errorf(
+			"route with id %s is already mapped. routes specificed in the 'routes' argument can only be mapped to one 'cf_app' resource",
```
I carried this restriction over from the original routes code.  I feel that it still makes sense because if you want to manage a route in this fashion, you should use the `cf_route` resource for this purpose. 
Furthermore, once we get into implementing blue/green support, the complexities of a proper blue/green deployment process in CF require that the `cf_app` resource is aware of all routes mapped to the application that need to continue to always route to a live application during the update, and those routes probably need to be exclusive to that application in order to make "blue/green sense".

gberche-orange replied:
@jcarrothers-sap 

It seems that this check is currently only performed on mapped routes during updates, as the following TF syntax at https://gist.github.com/gberche-orange/59d9075896e69313a145a44ae04e592c  indeed successfully resulted in mapping a given route to multiple apps with 0.9.9 version : 

``` 
cf a
Getting apps in org bercheg-org / space demo-basel as xx..
OK

name    requested state   instances   memory   disk   urls
dora1   started           1/1         1G       1G     dora-tf-demo.cfapps.io
dora2   started           1/1         1G       1G     dora-tf-demo.cfapps.io
```

The provider only failed after trying to further update app routes, as the restriction seems only applied when reading the existing route mapping. See related traces at https://gist.github.com/gberche-orange/c651aa2743faa6c958cae10b3ffe2524


> I carried this restriction over from the original routes code
>  I feel that it still makes sense because if you want to manage a route in this fashion, you should use the cf_route resource for this purpose.
> Furthermore, once we get into implementing blue/green support [...]

We'll have to see whether we indeed need blue/green support in the terraform provider, or whether a "zero downtime" is sufficient, see #25 and https://github.com/cloudfoundry/cloud_controller_ng/blob/master/docs/v3/zero_downtime_deployments.md

I still have preferences for:
1- removing this consistency check along with old route management
2- if we transiently keep it, document it in the `cloudfoundry_application` documentation

We can however live with this limitation if we don't find time to prioritize this, as we need to get 1.0 out asap. Feel free then to resolve this conversation if you feel we should tackle this post 1.0

jcarrothers-sap replied:
@gberche-orange 

>It seems that this check is currently only performed on mapped routes during updates, as the following TF syntax at https://gist.github.com/gberche-orange/59d9075896e69313a145a44ae04e592c  indeed successfully resulted in mapping a given route to multiple apps with 0.9.9 version : 

Based on an examination of the code, I believe that the reason your sample TF syntax works is due to a race condition as both applications would likely end up being created in parallel, but I am certain that the current create code does check and attempt to enforce the exclusivity of routes when they are mapped using the `cloudfoundry_app` resource as I have experienced this limitation in the past working with the older code base.

> We'll have to see whether we indeed need blue/green support in the terraform provider, or whether a "zero downtime" is sufficient, see #25 and https://github.com/cloudfoundry/cloud_controller_ng/blob/master/docs/v3/zero_downtime_deployments.md

It's possible the new "zero downtime" feature may be sufficient, but I do not believe it is yet available for my team's deployments of CF, and we are currently using an internal fork of the provider that has a preliminary, but working, implementation of the blue/green feature implemented in the provider.

> I still have preferences for:
> 1- removing this consistency check along with old route management

After thinking about this some more, I have realised that the exclusivity has more to do with the convention that Terraform takes full ownership of resources and is normally responsible for correcting configuration drift.  Without the need for blue/green this wouldn't be an issue as defining route mappings on the `cloudfoundry_app` resource wouldn't be implemented and we could just rely on using `cloudfoundry_route` resources to control the mappings. Then there wouldn't be any question about correcting configuration drift.  However, due to the thought that the provider was going to have to implement blue/green support on its own, this led to the need to be able to manage route mappings from the app resource.

> 2- if we transiently keep it, document it in the `cloudfoundry_application` documentation

I agree that this limitation should be documented.  

> We can however live with this limitation if we don't find time to prioritize this, as we need to get 1.0 out asap. Feel free then to resolve this conversation if you feel we should tackle this post 1.0

I propose that we discuss this in detail at the sync-up meeting on Monday.

jcarrothers-sap commented:
@gberche-orange I believe I have now addressed all of your review comments.

P.S. - I'm expecting many of the TestAccApp tests to fail due to a missing binary dependency.  I've contacted @mevansam about this problem, but we may need to find a more permanent solution to the issue of test failures when remote dependencies disappear.

jcarrothers-sap commented:
145de4c should fix the broken tests.

janosbinder commented:
LGTM - please merge this PR

gberche-orange commented:
Agreed with @jcarrothers-sap to document in a distinct PR the limitation that a route can only be bound to a single application, along with hint about not using cloudfoundry_route.target.app along with cloudfoundry_application.route. Would also include fix for [cloudfoundry_route.target.app](https://github.com/mevansam/terraform-provider-cf/blob/dev/website/docs/r/route.html.markdown) whose type should reference a set instead of a string (i.e. support multiple target apps)
